File: wallet-kit/src/main/java/bitcoin/walllet/kit/network/message/BlockMessage.java
package bitcoin.walllet.kit.network.message;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Arrays;

import bitcoin.walllet.kit.common.io.BitcoinInput;
import bitcoin.walllet.kit.struct.Block;

/**
 * Build P2P message:
 * https://en.bitcoin.it/wiki/Protocol_documentation#Message_structure
 *
 * @author liaoxuefeng
 */
public class BlockMessage extends Message {

    static final Logger log = LoggerFactory.getLogger(BlockMessage.class);

    public Block block;

    public BlockMessage() {
        super("block");
    }

    public BlockMessage(byte[] payload) throws IOException {
        super("block");
        try (BitcoinInput input = new BitcoinInput(new ByteArrayInputStream(payload))) {
            this.block = new Block(input);
        }
    }

    @Override
    protected byte[] getPayload() {
        return this.block.toByteArray();
    }

    /**
     * Validate block hash.
     */
    public boolean validateHash() {
        byte[] merkleHash = this.block.calculateMerkleHash();
        if (!Arrays.equals(merkleHash, this.block.header.merkleHash)) {
            log.error("Validate merckle hash failed.");
            return false;
        }
        // TODO: validate bits:
        return true;
    }

    @Override
    public String toString() {
        return "BlockMessage(txnCount=" + this.block.txns.length + ")";
    }

}

File: wallet-kit/src/main/kotlin/bitcoin/wallet/kit/messages/BlockMessage.kt
package bitcoin.wallet.kit.messages

import bitcoin.wallet.kit.models.Header
import bitcoin.wallet.kit.models.Transaction
import bitcoin.walllet.kit.io.BitcoinInput
import bitcoin.walllet.kit.io.BitcoinOutput
import java.io.ByteArrayInputStream
import java.io.IOException

/**
 * The 'block' message consists of a single serialized block.
 */
class BlockMessage() : Message("block") {

    lateinit var header: Header
    lateinit var transactions: Array<Transaction>

    @Throws(IOException::class)
    constructor(payload: ByteArray) : this() {
        BitcoinInput(ByteArrayInputStream(payload)).use { input ->
            header = Header(input)
            val txCount = input.readVarInt() // do not store count
            transactions = Array(txCount.toInt()) {
                Transaction(input)
            }
        }
    }

    override fun getPayload(): ByteArray {
        val output = BitcoinOutput()
        output.write(header.toByteArray())
        output.writeVarInt(transactions.size.toLong())
        for (transaction in transactions) {
            output.write(transaction.toByteArray())
        }

        return output.toByteArray()
    }

    /**
     * Validate block hash.
     */
    fun validateHash(): Boolean {
        // TODO: validate bits:
        return true
    }

    override fun toString(): String {
        return "BlockMessage(txCount=${transactions.size})"
    }
}
